src/suricata-plugin.h

```diff
@@ -41,16 +41,30 @@ typedef struct SCPlugin_ {
     void (*Init)(void);
 } SCPlugin;
 
+typedef struct SCPluginFileOptions_ {
+    bool threaded;
+    const char *filename; /* Configured file name */
+    const char *log_dir;
+    const char *append;
+} SCPluginFileOptions;
+
 /**
  * Structure used to define a file type plugin.
  *
  * Currently only used by the Eve output type.
  */
 typedef struct SCPluginFileType_ {
     char *name;
-    int (*Open)(ConfNode *conf, void **data);
-    int (*Write)(const char *buffer, int buffer_len, void *ctx);
+    /* Called on first open; no other calls will be made if unthreaded */
+    int (*Open)(ConfNode *conf, SCPluginFileOptions *opt, void **ctx);
+    /* Write - Called on each write to the object */
+    int (*Write)(const char *buffer, int buffer_len, void *ctx, void *thread_ctx);
+    /* Close - Called on final close */
     void (*Close)(void *ctx);
+    /* Open - Called for each thread using the file object */
+    int (*ThreadOpenFunc)(void *ctx, int thread_id, void **thread_ctx);
+    /* Close - Called for each thread using the file object */
+    int (*ThreadCloseFunc)(void *ctx, void *thread_ctx);
```

Comment:
The naming still confuses me...  Maybe its the multiple opens..
* Open -> Init: as its a one time initializer, could open files...
* ThreadOpenFunc -> ThreadInit

Then we document... If not threaded, you will get a one time call to Init.. Open your file, database connection socket, etc. here... If threading is enabled this is where you will get your configuration data, another call for each thread will be made the ThreadInit where you can open files, databases, sockets, whatever..